### PR TITLE
raise KeyError if registering buffer/param when attr exists

### PR DESCRIPTION
**Pull request by alykhantejani**
Fix for #1874

### CODE REVIEW

**Comment by alykhantejani** [comment]
(not quite fixed - working on it)

**Review by apaszke — COMMENTED** [review]
Won't this snippet raise an error after this PR?
```python
class MyModule(nn.Module):
    def __init__(self):
        super(MyModule, self).__init__()
        self.register_parameter('x', ...)
        ...
        self.register_parameter('x', ...)
```

I'm not sure about this PR. The way I've been thinking about these functions is that they're pretty much interchangeable with `module.param = ...`, which obviously doesn't raise an error when you do it twice. But maybe we should make them a way to enforce additional checks 😕

**Comment by alykhantejani** [comment]
@apaszke yeah, the more I'm looking into fixing the broken tests, the more complicated it's becoming. 

I thought it'd be something quick before I went to bed, I'm now well and truly down the rabbit hole.

**Comment by alykhantejani** [comment]
@apaszke The more I think about it, it's very pythonic to allow the user to just to do
```python
m.a = Parameter(...)
m.a = nn.Linear(...)
```
But we don't want them to do:
```Python
m.a = Parameter(...)
m.add_module('a', nn.Linear(...))
```
So if we want the `register_parameter`, `register_buffer` and `add_module` functions to behave like `__setattr__`, then we can leave them as is. Otherwise I think we should change both the functions and `__setattr__` to not allow overwriting of already registered buffers/params, and you should explicitly `del` them first if you want to overwrite. 

Not sure which is the best option though?

Anyway, its 2am here and I've left my charger at the office, let me know what your thoughts are - I'm happy to adjust the PR accordingly or alternatively close if we mark this as "wont fix"

**Comment by apaszke** [comment]
I think we can make them safer alternatives to regular assignments. Let's add the checks

**Review by apaszke — DISMISSED** [review]
Looks good, but they should all be consistent - allow these functions to overwrite existing attributes of the same kind (buffers can override buffers, params params, etc.), but not otherwise. Currently only `register_parameter` allows that. Once this is implemented this is ready to go

**Comment by alykhantejani** [comment]
@apaszke sorry for leaving this stale for a while - I'll try to get to it in the next week

**Comment by alykhantejani** [comment]
@apaszke I think this is now ready - let me know what you think

File: torch/nn/modules/module.py [hunk 1]
```diff
@@ -76,6 +76,9 @@ def register_buffer(self, name, tensor):
         Example:
             >>> self.register_buffer('running_mean', torch.zeros(num_features))
         """
+        if hasattr(self, name) and name not in self._buffers.keys():
```
Comment (fmassa) [inline torch/nn/modules/module.py]: Do you need the `.keys()` here? Can't you just do instead `and name not in self._buffers`? Same for the other occurences.

**Review by fmassa — COMMENTED** [review]
This looks good to me. I have only one minor comment. Thanks Aly!

File: torch/nn/modules/module.py (hunk 1 above)
Comment (alykhantejani) [reply torch/nn/modules/module.py]: Didn't know you could even do that 👍 - will change that and push

**Comment by soumith** [comment]
Thanks Aly!

**Pull request merged**